### DISCUSSION TITLE
How to add a self-defined state variable in moose source codes

### DISCUSSION THREAD

**Discussion by pshen20127** [idaholab/moose#23182] [Q&A Modules: Solid mechanics] [answered]
Hello all

I want to add a internal state variable in the file 'CrystalPlasticityKalidindiUpdate.C'. But I met this error message as below.  What should I do to fix it? The 'shearstrain' in error message is the state variable I defined.

application called MPI_Abort(MPI_COMM_WORLD, 1) - process 3

*** ERROR ***
The following error occurred in the object "MOOSE Problem", of type "FEProblem".

Material property 'shearstrain', requested by 'shearstrain_0' is not defined on block GRAIN105_SET_QUAD4
Material property 'shearstrain', requested by 'shearstrain_1' is not defined on block GRAIN105_SET_QUAD4
Material property 'shearstrain', requested by 'shearstrain_10' is not defined on block GRAIN105_SET_QUAD4
Material property 'shearstrain', requested by 'shearstrain_11' is not defined on block GRAIN105_SET_QUAD4
Material property 'shearstrain', requested by 'shearstrain_2' is not defined on block GRAIN105_SET_QUAD4
Material property 'shearstrain', requested by 'shearstrain_3' is not defined on block GRAIN105_SET_QUAD4
Material property 'shearstrain', requested by 'shearstrain_4' is not defined on block GRAIN105_SET_QUAD4
Material property 'shearstrain', requested by 'shearstrain_5' is not defined on block GRAIN105_SET_QUAD4

Thanks
pshen20127

**Comment by GiudGiud** [comment, marked as answer]
Hello

You need to make sure this property `shearstrain` is being defined on the relevant block.
How is it being defined currently?

Guillaume

**Reply by pshen20127** [reply]
Thanks for your reply. I attached modified source codes and input file. You can search for 'shearstrain' in these files. In input file, all shearstrain-related settings follow the 'slip increment'. I am not sure if I should use type = MaterialStdVectorAux for shearstrain in AuxKernels.

[modified source codes.zip](https://github.com/idaholab/moose/files/10439850/modified.source.codes.zip)

**Reply by GiudGiud** [reply]
We try to avoid downloading zip files from the internet. Can you please make a github repo?

I ll try to check it out next week.

**Reply by pshen20127** [reply]
Thank you so much. please see the link: https://github.com/pshen20127/IN625.git
Forgot to mention that when I modified the source codes, I copied the orginal files and change the filename and corresponding classes in codes by adding a suffix 'AddStateVar'. So don't be surprised these filenames are different from original ones.

**Reply by GiudGiud** [reply]
Shearstrain has been declared like this
```
    _shearstrain(declareProperty<std::vector<Real>>(_base_name + "_shearstrain")),
```

so it s name is never shearstrain.
It's either base_name_shearstrain OR if base_name is not used in the input file it's `_shearstrain` with the underscore

**Reply by pshen20127** [reply]
Thank you so much. I fix it. What is the '_base_name' in

`_shearstrain(declareProperty<std::vector<Real>>(_base_name + "_shearstrain")),`

I don't know where it is used.

**Reply by GiudGiud** [reply]
it's to prefix material properties if you had multiple phases in the grains or something. Let's you define the same type of properties multiple times

**Reply by pshen20127** [reply]
Got it. Do you know how to define type in Auxkernels and postprocessor? I dont' know how many do they have and how to choose.
>   [shearstrain_1]
   type = MaterialStdVectorAux
   variable = shearstrain_1
   property = shearstrain
   index = 1
   execute_on = timestep_end
  []
  [shearstrain_1]
    type = ElementAverageValue
    variable = shearstrain_1
  []

**Reply by GiudGiud** [reply]
If you just want to output the material properties to exodus you should use the 'outputs' parameter of the material. 

Otherwise you will need to define many auxkernels and many posptprocessors for all the ones you care about

**Reply by pshen20127** [reply]
Is there a file in source codes recording which material properties I can output? Or where I can learn this content? 

If I define a new variable,  how do I set types in auxkernels and postprocessors.  I'm always struggled with this part.

**Reply by GiudGiud** [reply]
for any material you can specify material outputs parameters. Look at the list of parameters of any materials.

**Reply by pshen20127** [reply]
Got it, thank you.
